src/graph.rs

```diff
@@ -0,0 +1,924 @@
+
+    /// Generate active tasks graph.
+    pub(crate) fn get_tasks_per_second_graph(&self) -> Graph<usize> {
+        Graph::new(
+            "graph-tps",
+            "Tasks #",
+            self.add_timestamp_to_html_graph_data(&self.tasks_per_second),
+            self.starting,
+            self.started,
+            self.stopping,
+            self.stopped,
+        )
+    }
+
+    /// Generate errors per second graph.
+    pub(crate) fn get_errors_per_second_graph(&self) -> Graph<u32> {
+        Graph::new(
+            "graph-eps",
+            "Errors #",
+            self.add_timestamp_to_html_graph_data(&self.errors_per_second),
+            self.starting,
+            self.started,
+            self.stopping,
+            self.stopped,
+        )
+    }
+
+    /// Adds timestamps to the graph data series to ensure correct time display on x axis.
+    ///
+    /// Will take a vector of (generally numerical) values and convert them into tuples where
+    /// the second element will be the data point and the first element will be formatted time
+    /// it belongs to.
+    fn add_timestamp_to_html_graph_data<T: Copy>(&self, data: &[T]) -> Vec<(String, T)> {
+        // TODO properly handle the date Option timestamps.
+        data.iter()
+            .enumerate()
+            .filter(|(second, _)| {
+                if self.include_startup {
+                    true
+                } else {
+                    *second as i64 + self.starting.unwrap().timestamp()
+                        >= self.started.unwrap().timestamp()
+                }
+            })
+            .map(|(second, &count)| {
+                (
+                    Local
+                        .timestamp(second as i64 + self.starting.unwrap().timestamp(), 0)
+                        .format("%Y-%m-%d %H:%M:%S")
+                        .to_string(),
+                    count,
+                )
+            })
+            .collect::<Vec<_>>()
+    }
+}
+
+/// Expands vectors that collect per-second data for HTML report graphs with a
+/// default value.
+///
+/// We need to do that since we don't know for how long the load test will run
+/// and we can't initialize these vectors at the beginning. It is also
+/// better to do it as we go to save memory.
+fn expand_per_second_metric_array<T: Clone>(data: &mut Vec<T>, second: usize, initial: T) {
+    // Each element in per second metric vectors (self.requests_per_second,
+    // self.errors_per_second, ...) is counted for a given second since the start
+    // of the test. Since we don't know how long the test will at the beginning
+    // we need to push new elements (second counters) as the test is running.
+    if data.len() <= second {
+        for _ in 0..(second - data.len() + 1) {
+            data.push(initial.clone());
+        }
+    };
+}
+
+/// Defines the HTML graph data.
+#[derive(Debug)]
+// TODO why does this need to be pub (instead pub(crate)) in order for stuff on
+// report.rs to not complain?
```

Comment:
Typically this would be because the structure it's stored in is `pub`, not `pub(crate)`, and therefore requires everything it contains to also be `pub`.

Reply:
Thanks, makes sense! I changed the functions in `report.rs` to be `pub(crate)`.